File: headfirst/03_decorator/StarBuzzCoffee.java
abstract class Beverage {
    String description = "Unknown Beverage";

    String getDescription() {
        return description;
    }

    abstract double cost();
}

abstract class CondimentDecorator extends Beverage {
    abstract String getDescription();
}

//_________________________________ Some Kinds of Beverage ____________________________________
class Espresso extends Beverage {
    Espresso() {
        description = "Espresso";
    }

    double cost() {
        return 1.99;
    }
}

class HouseBlend extends Beverage {
    HouseBlend() {
        description = "House Blend Coffee";
    }

    double cost() {
        return .89;
    }
}

//_____________________________________ Some Condiment ____________________________________________

class Mocha extends CondimentDecorator {
    Beverage beverage;

    Mocha(Beverage beverage){
        this.beverage = beverage;
    }

    String getDescription() {
        return beverage.getDescription() + ", Mocha";
    }

    double cost() {
        return .20 + beverage.cost();
    }
}

class Soy extends CondimentDecorator {
    Beverage beverage;

    Soy(Beverage beverage) {
        this.beverage = beverage;
    }

    String getDescription() {
        return beverage.getDescription() + ", Soy";
    }

    double cost() {
        return .15 + beverage.cost();
    }
}

class Whip extends CondimentDecorator {
    Beverage beverage;

    Whip(Beverage beverage) {
        this.beverage = beverage;
    }

    String getDescription() {
        return beverage.getDescription() + ", Whip";
    }

    double cost() {
        return .10 + beverage.cost();
    }
}

//______________________________________ StarBuzz Coffee ____________________________________________

public class StarBuzzCoffee {
    public static void main(String[] args) {
        Beverage beverage = new Espresso();
        System.out.println(beverage.getDescription() + " $ " + beverage.cost());
        
        Beverage beverage2 = new HouseBlend();
        beverage2 = new Mocha(beverage2);
        beverage2 = new Mocha(beverage2);
        beverage2 = new Whip(beverage2);
        System.out.println(beverage2.getDescription() + " $ " + beverage2.cost());
        
        Beverage beverage3 = new Espresso();
        beverage3 = new Soy(beverage3);
        beverage3 = new Mocha(beverage3);
        beverage3 = new Whip(beverage3);
        System.out.println(beverage3.getDescription() + " $ " + beverage3.cost());
    }
}

File: headfirst/03_decorator/starbuzz_coffee.go
package main

import (
	"fmt"
)

const (
	Mocha CondimentType = "Mocha"
	Soy   CondimentType = "Soy"
	Whip  CondimentType = "Whip"
)

const (
	Espresso   CoffeeType = "Espresso"
	HouseBlend CoffeeType = "House Blend Coffee"
)

var CondimentPriceTable = map[CondimentType]float32{
	Mocha: .20,
	Soy:   .15,
	Whip:  .10,
}

var CoffeePriceTable = map[CoffeeType]float32{
	Espresso:   1.99,
	HouseBlend: .89,
}

type CondimentType string
type CoffeeType string

// ____________________ Condiment _________________________

type Condiment struct {
	Type  CondimentType
	Price float32
}

func NewCondiment(typ CondimentType) *Condiment {
	return &Condiment{
		Type:  typ,
		Price: CondimentPriceTable[typ],
	}
}

//_____________________ Some Kind of Coffee ____________________

type Coffee struct {
	Type       CoffeeType
	Price      float32
	Condiments []*Condiment
}

func NewCoffee(typ CoffeeType) *Coffee {
	return &Coffee{
		Type:       typ,
		Price:      CoffeePriceTable[typ],
		Condiments: make([]*Condiment, 0),
	}
}

func (c *Coffee) Add(cdm *Condiment) {
	c.Condiments = append(c.Condiments, cdm)
}

func (c *Coffee) Cost() float32 {
	var price float32
	for _, cdm := range c.Condiments {
		price += cdm.Price
	}
	return price + c.Price
}

func (c *Coffee) GetDescription() string {
	var decorates string
	for _, cdm := range c.Condiments {
		decorates += string(cdm.Type) + ", "
	}
	return decorates + string(c.Type)
}

func main() {
	beverage := NewCoffee(Espresso)
	fmt.Printf("Desc: %s, Price: %.2f $\n\n", beverage.GetDescription(), beverage.Cost())

	beverage2 := NewCoffee(HouseBlend)
	beverage2.Add(NewCondiment(Mocha))
	beverage2.Add(NewCondiment(Mocha))
	beverage2.Add(NewCondiment(Whip))
	fmt.Printf("Desc: %s, Price: %.2f $\n\n", beverage2.GetDescription(), beverage2.Cost())

	beverage3 := NewCoffee(Espresso)
	beverage3.Add(NewCondiment(Soy))
	beverage3.Add(NewCondiment(Mocha))
	beverage3.Add(NewCondiment(Whip))
	fmt.Printf("Desc: %s, Price: %.2f $\n\n", beverage3.GetDescription(), beverage3.Cost())
}
